hwx_malloc.c

```diff
@@ -1,36 +1,360 @@
+							curr_node += size;
+							curr_node->size = new_free;
+							curr_node->next = temp.next;
+							prev_node->next = curr_node;
+
+						}
+						// no new free memory to add
+						new_free = 0;
+
+						// break from loop
+						break;
+
+					}
+					// check if there is exactly enough space
+					else if (curr_node->size + sizeof(node_t) == size) {
+						// remove current node from free list
+						prev_node->next = curr_node->next;
+
+						// reallocated freed memory
+						alloced = curr_node;
+
+						// no new free memory
+						new_free = 0;
+
+						// break from loop
+						break;
+
+					}
+
+					// if no free heap space allocate new page
+					else if (curr_node->next == NULL){
+
+						// initially begin with one page
+						pages = 1;
+
+						// check if there will be any overhead
+						if (((pages * PAGE_SIZE) - size) && ((pages * PAGE_SIZE) - size) < (sizeof(node_t) + 1))
+							pages++;
+						
+						// increment stats
+						stats.pages_mapped += pages;
+
+						// allocate new page
+						alloced = mmap(NULL, 
+											 		(PAGE_SIZE * pages),
+											 		(PROT_READ | PROT_WRITE),
+											 		(MAP_ANON | MAP_PRIVATE),
+											 		-1, 0);
+
+						// increment unmap flag
+						MUNMAP += 1;
+						
+						// calculate total amount of bytes mmapped
+						total_mapped = (pages * PAGE_SIZE);
+			
+						// calculate amount of free memory mapped
+						new_free = (total_mapped - size);
+	
+						break;
+					}	
+					// save previous node
+					prev_node = curr_node;
+
+				} while (curr_node = curr_node->next);
+			}
+
+		// add header to alloced memory
+		header_alloced = (header_t*) alloced;
+		header_alloced->size = (size - sizeof(header_t));
+
+		// increment alloced pointer by header size
+		alloced += sizeof(header_t);
+
+		// add any unallocated but mapped data to free list
+		if (new_free && size < PAGE_SIZE) {
+
+			// find beginning of free memory
+			free_mem = (node_t*) (((void*) header_alloced) + size);
+
+			// subtract mapping overhead
+			new_free -= sizeof(node_t);
+			free_mem->size = new_free;
+
+			// if free list is empty add free memory to it
+			if (FREE_LIST == NULL) {
+					free_mem->next = NULL;
+					FREE_LIST = free_mem;
+
+			}
+			// if not first eleement add to free list
+			else {
+				_add_to_free_list(FREE_LIST, free_mem, 0);
+
+			}
+		}
+		// return allocated memory pointer
+		return alloced;
+}
+
+void
+xfree(void* item)
 {
-    // TODO: write realloc
-    return 0;
+    stats.chunks_freed += 1;
+
+    // Actually free the items
+		node_t *curr_node;
+		node_t *free_node;
+		size_t block_size;
+
+		// size of memory block to free
+		block_size = *((size_t*) (item - (sizeof(header_t))));
+
+		if ((block_size) >= PAGE_SIZE - sizeof(header_t)) {
+			// immediately munmap any allocation greater than or equal to page
+			stats.pages_unmapped += div_up(block_size, PAGE_SIZE - sizeof(header_t));
+			munmap(item - sizeof(header_t), block_size);
```

Comment:
So is this where we would add another check to optomize mumapping, or is that handled because you munmap during coalessing?

Reply:
Yeah im pretty sure